### PR TITLE
support DINO algo

### PR DESCRIPTION
**Pull request by tuofeilunhifi**
dino_4sc_r50_12e：48.71
dino_4sc_r50_24e：50.53
dino_4sc_r50_36e：50.69
dino_4sc_swinl_12e: 56.86
dino_4sc_swinl_36e: 58.04
dino_5sc_swinl_36e: 58.47

### CODE REVIEW

File: easycv/datasets/loader/build_loader.py [hunk 1]
```diff
@@ -145,12 +153,13 @@ def build_dataloader(dataset,
...
-    if seed is not None:
-        worker_seed = worker_id + seed
-        np.random.seed(worker_seed)
-        random.seed(worker_seed)
-        torch.manual_seed(worker_seed)
+def worker_init_fn(worker_id, num_workers, rank, seed, odps_config=None):
+    # The seed of each worker equals to
+    # num_worker * rank + worker_id + user_seed
+    worker_seed = num_workers * rank + worker_id + seed
+    np.random.seed(worker_seed)
```
Comment (wuziheng) [inline easycv/datasets/loader/build_loader.py:160]: 小问一下，这个地方的默认行为从 不给seed 就不设置seed ，变成一定会随机按照worker_id 设置seed 对么

Comment (tuofeilunhifi) [reply easycv/datasets/loader/build_loader.py:160]: 对的，就是要保证所有部分的随机种子由一个seed控制，要不然会影响精度，大概能差0.1-0.2之间https://github.com/open-mmlab/mmdetection/pull/8362

Comment (Li-Qingyun) [reply easycv/datasets/loader/build_loader.py:160]: 这个种子怎么设会影响精度么，我看源码里是 seed + rank 的设法，没有 works 这一项呀

Comment (tuofeilunhifi) [reply easycv/datasets/loader/build_loader.py:160]: > 这个种子怎么设会影响精度么，我看源码里是 seed + rank 的设法，没有 wo

这个worker_id没有检查过，不知道是否有影响

Comment (wuziheng) [reply easycv/datasets/loader/build_loader.py:160]: 1. 所以我的问题是，如果说这个 seed 很影响某一个算法的精度，那么我们应该保持说整体行为不变的情况下，针对这个算法的build dataloader 的初始化的seed 可以做到指定，从而保证算法行为的正确。

当然根据 line62， 确实做了对以前的兼容，对于没有输入seed的算法，引入mmcv 的最新的init_random_seed 去保证不同rank 之间的 seed是相同的，但以前的行为是所有的rank是自行获取seed（不做行为的默认），大概率不会相同。

这里的问题就是
1. 这么大的行为变动（影响所有算法在分布式情况下的random seed）在62行没有任何说明是不利于其他collaborate 去发现这一改动的。所有人会默认行为没有变动只是改变了实现。
2. 对于其他的算法，是不是都需要所有rank 保持同样的num seed，有没有做过前向兼容的测试和复现，如果没有，这么大的改动的原因在哪里？仅仅是因为mmdetection 的comments some potential bug ？

File: easycv/models/segmentation/heads/pixel_decoder.py [hunk 2]
```diff
@@ -11,7 +11,7 @@
 from .transformer_decoder import PositionEmbeddingSine, _get_activation_fn
 
 try:
-    from thirdparty.deformable_transformer.modules import MSDeformAttn
+    from thirdparty.msdeformattn.modules import MSDeformAttn
 except:
     pass
```
Comment (wuziheng) [inline easycv/models/segmentation/heads/pixel_decoder.py:16]: try.... except....pass with no logging or raise 是不推荐使用的用法，希望这里还是能加一个报错或者说明影响的范围，比如我这种不明白为什么要import 这个东西的读者 会质疑，如果可以try except pass with no logging ，那为什么不直接删掉。

File: easycv/models/utils/transformer.py [hunk 3]
```diff
@@ -22,6 +23,60 @@ def forward(self, x):
         return x
 
 
+class Mlp(nn.Module):
+    """ Multilayer perceptron."""
+
+    def __init__(self,
+                 in_features,
```
Comment (wuziheng) [inline easycv/models/utils/transformer.py:30]: add help doc for nn.module init

File: easycv/datasets/loader/build_loader.py (hunk 1 above)
Comment (tuofeilunhifi) [reply easycv/datasets/loader/build_loader.py:160]: 有别人的对比实验，不同rank使用不同seed会降低精度，https://github.com/open-mmlab/mmdetection/pull/8362

Comment (tuofeilunhifi) [reply easycv/datasets/loader/build_loader.py:160]: 可以加一个--diff-seed参数进行控制

File: easycv/models/segmentation/heads/pixel_decoder.py (hunk 2 above)
Comment (tuofeilunhifi) [reply easycv/models/segmentation/heads/pixel_decoder.py:16]: > try.... except....pass with no logging or raise 是不推荐使用的用法，希望这里还是能加一个报错或者说明影响的范围，比如我这种不明白为什么要import 这个东西的读者 会质疑，如果可以try except pass with no logging ，那为什么不直接删掉。

这里之前贺弘可能因为msdeformattn不是所有算法必须的，保证不会报错吧

File: easycv/datasets/loader/build_loader.py (hunk 1 above)
Comment (wuziheng) [reply easycv/datasets/loader/build_loader.py:160]: 1. 你给的 链接还是关于dino的，但你的修改是针对所有算法的，并不是不认同关于这个算法的修改，而是你的修改还会影响到其他的，而且似乎你没有在意这一点，所以即使加上一个diff-seed，这一类的改动也需要在PR里明确指出来。否则还是存在风险的。
2.  关于dino的这个seed的link我觉得可以放在代码修改处的链接上。

File: easycv/models/utils/transformer.py (hunk 3 above)
Comment (tuofeilunhifi) [reply easycv/models/utils/transformer.py:33]: done

File: easycv/models/segmentation/heads/pixel_decoder.py (hunk 2 above)
Comment (tuofeilunhifi) [reply easycv/models/segmentation/heads/pixel_decoder.py]: done

File: configs/detection/dino/dino_4sc_r50_12e_coco.py [hunk 4]
```diff
@@ -0,0 +1,4 @@
+_base_ = [
```
Comment (wenmengzhou) [inline configs/detection/dino/dino_4sc_r50_12e_coco.py:1]: add README.md for dino

File: easycv/utils/dist_utils.py [hunk 5]
```diff
@@ -134,3 +136,61 @@ def all_reduce_dict(py_dict, op='sum', group=None, to_float=True):
...
+
+    if rank == 0:
+        random_num = torch.tensor(seed, dtype=torch.int32, device=device)
+    else:
+        random_num = torch.tensor(0, dtype=torch.int32, device=device)
+    dist.broadcast(random_num, src=0)
+    return random_num.item()
+
+
+def set_random_seed(seed, deterministic=False):
```
Comment (wenmengzhou) [inline easycv/utils/dist_utils.py:178]: this is not a util function for distributed training, seems not appropriate to be here

File: easycv/hooks/optimizer_hook.py [hunk 6]
```diff
@@ -48,9 +46,6 @@ def __init__(self,
         self.multiply_key = multiply_key
         self.multiply_rate = multiply_rate
 
-    def before_run(self, runner):
```
Comment (tuofeilunhifi) [inline easycv/hooks/optimizer_hook.py:51]: remove before_run zero_grad()，because original MMCV was not used. dino experiment is remove zero_grad() https://github.com/open-mmlab/mmcv/blob/master/mmcv/runner/hooks/optimizer.py

File: easycv/models/segmentation/heads/pixel_decoder.py (hunk 2 above)
Comment (wenmengzhou) [reply easycv/models/segmentation/heads/pixel_decoder.py]: delay the following import to where MSDeformAttn is used but not at the top of file

from thirdparty.msdeformattn.modules import MSDeformAttn

File: easycv/models/backbones/vision_transformer.py [hunk 7]
```diff
@@ -635,7 +595,7 @@ def forward(self, x):
 
 # todo: refactor vitdet and vit_transformer_dynamic
 @BACKBONES.register_module()
-class ViTDet(nn.Module):
```
Comment (wenmengzhou) [inline easycv/models/backbones/vision_transformer.py:638]: keep original name until we merge those vision transformers into one.  This vit is adapted with support for vitdet

File: easycv/utils/dist_utils.py (hunk 5 above)
Comment (tuofeilunhifi) [reply easycv/utils/dist_utils.py]: done

File: easycv/models/segmentation/heads/pixel_decoder.py (hunk 2 above)
Comment (tuofeilunhifi) [reply easycv/models/segmentation/heads/pixel_decoder.py]: done

File: easycv/models/backbones/vision_transformer.py (hunk 7 above)
Comment (tuofeilunhifi) [reply easycv/models/backbones/vision_transformer.py]: done

File: easycv/datasets/loader/build_loader.py (hunk 1 above)
Comment (tuofeilunhifi) [reply easycv/datasets/loader/build_loader.py:160]: done

File: configs/detection/dino/dino_4sc_r50_12e_coco.py (hunk 4 above)
Comment (tuofeilunhifi) [reply configs/detection/dino/dino_4sc_r50_12e_coco.py:1]: done

File: easycv/models/detection/utils/misc.py [hunk 8]
```diff
@@ -169,21 +174,173 @@ def output_postprocess(outputs, img_metas=None):
...
+                            dim=3).flatten(2)
+
+        pos = torch.cat((pos_y, pos_x, pos_w, pos_h), dim=2)
+    else:
+        raise ValueError('Unknown pos_tensor shape(-1):{}'.format(
+            pos_tensor.size(-1)))
+    return pos
+
+
+class DetrPostProcess(nn.Module):
```
Comment (Cathy0908) [inline easycv/models/detection/utils/misc.py:292]: move DetrPostProcess to detr_utils.py

File: easycv/models/detection/utils/boxes.py [hunk 9]
```diff
@@ -511,3 +509,11 @@ def batched_nms(boxes, scores, idxs, nms_cfg, class_agnostic=False):
 
     boxes = torch.cat([boxes, scores[:, None]], -1)
     return boxes, keep
+
+
+def fp16_clamp(x, min=None, max=None):
```
Comment (Cathy0908) [inline easycv/models/detection/utils/boxes.py:514]: 这个接口和box有啥关系，为什么放在boxes.py里

File: easycv/models/detection/utils/set_criterion.py [hunk 10]
```diff
@@ -228,6 +245,120 @@ def forward(self, outputs, targets, mask_dict=None, return_indices=False):
         return losses
 
 
+class CDNCriterion(SetCriterion):
```
Comment (Cathy0908) [inline easycv/models/detection/utils/set_criterion.py:248]: 这个文件里都是loss的接口，移动到loss模块里

File: thirdparty/msdeformattn/modules/ms_deform_attn.py [hunk 11]
```diff
@@ -32,7 +32,7 @@ def _is_power_of_2(n):
 
 class MSDeformAttn(nn.Module):
```
Comment (Cathy0908) [inline thirdparty/msdeformattn/modules/ms_deform_attn.py:33]: thirdparty/msdeformattn 改为 thirdparty/deformable_attention

File: docs/source/model_zoo_det.md [hunk 12]
```diff
@@ -35,3 +37,13 @@ Pretrained on COCO2017 dataset.
...
 | DN-DETR-r50    | [dab-detr-r50](https://github.com/alibaba/EasyCV/tree/master/configs/detection/dab_detr/dn_detr_r50_8x2_50e_coco.py) | 23M/43M | 58.5ms | 44.39                   | 64.66          | [model](https://pai-vision-data-hz.oss-cn-zhangjiakou.aliyuncs.com/EasyCV/modelzoo/detection/dn_detr/dn_detr_epoch_50.pth) - [log](https://pai-vision-data-hz.oss-cn-zhangjiakou.aliyuncs.com/EasyCV/modelzoo/detection/dn_detr/20220713_105127.log.json) |
+
+## DINO
+
+| Algorithm  | Config                                                       | Params<br/>(backbone/total)                            | inference time(V100)<br/>(ms/img)                      | bbox_mAP<sup>val<br/><sub>0.5:0.95</sub> | AP<sup>val<br/><sub>50</sub> | Download                                                     |    Comment         |
```
[the remark this answers is not in the document]
Comment (tuofeilunhifi) [reply docs/source/model_zoo_det.md:43]: done

File: docs/source/model_zoo_det.md [hunk 13]
```diff
@@ -18,9 +20,9 @@ Pretrained on COCO2017 dataset.
...
+| ---------- | ------------------------------------------------------------ | ------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ | --------------------------------------------- |
+| ViTDet_MaskRCNN    | [vitdet_maskrcnn](https://github.com/alibaba/EasyCV/tree/master/configs/detection/vitdet/vitdet_100e.py) | 88M/118M | 163ms | 50.57                   | 44.96          | [model](https://pai-vision-data-hz.oss-cn-zhangjiakou.aliyuncs.com/EasyCV/modelzoo/detection/vitdet/vit_base/vitdet_maskrcnn.pth) - [log](https://pai-vision-data-hz.oss-cn-zhangjiakou.aliyuncs.com/EasyCV/modelzoo/detection/vitdet/vit_base/vitdet_maskrcnn.log.json) |推理的输入尺寸为1024x1024|
```
[the remark this answers is not in the document]
Comment (tuofeilunhifi) [reply docs/source/model_zoo_det.md]: done

File: docs/source/model_zoo_det.md [hunk 14]
```diff
@@ -1,5 +1,7 @@
 # Detection Model Zoo
 
+(ps: 推理的输入尺寸默认为1333x800，机器默认为V100 16G)
```
[the remark this answers is not in the document]
Comment (tuofeilunhifi) [reply docs/source/model_zoo_det.md]: done

File: configs/detection/dino/README.md [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+
+<div align=center>
+<img src="https://pai-vision-data-hz.oss-cn-zhangjiakou.aliyuncs.com/EasyCV/algo_images/detection/DINO.png"/>
+</div>
+
+## Results and Models
+
+| Algorithm  | Config                                                       | Params<br/>(backbone/total)                            | inference time(V100)<br/>(ms/img)                      | bbox_mAP<sup>val<br/><sub>0.5:0.95</sub> | AP<sup>val<br/><sub>50</sub> | Download                                                     |    Comment         |
```
[the remark this answers is not in the document]
Comment (tuofeilunhifi) [reply configs/detection/dino/README.md]: done

File: easycv/models/detection/utils/misc.py (hunk 8 above)
Comment (tuofeilunhifi) [reply easycv/models/detection/utils/misc.py]: done

File: thirdparty/msdeformattn/modules/ms_deform_attn.py (hunk 11 above)
Comment (tuofeilunhifi) [reply thirdparty/msdeformattn/modules/ms_deform_attn.py]: done

File: easycv/models/detection/utils/set_criterion.py (hunk 10 above)
Comment (tuofeilunhifi) [reply easycv/models/detection/utils/set_criterion.py:248]: done

File: easycv/models/detection/utils/boxes.py (hunk 9 above)
Comment (tuofeilunhifi) [reply easycv/models/detection/utils/boxes.py]: done

File: configs/detection/common/dataset/autoaug_coco_detection.py [hunk 16]
```diff
@@ -15,44 +15,49 @@
 ]
 
 # dataset settings
-data_root = 'data/coco/'
+data_root = '/root/data/coco/'
```
Comment (wenmengzhou) [inline configs/detection/common/dataset/autoaug_coco_detection.py:18]: data/coco/

Comment (tuofeilunhifi) [reply configs/detection/common/dataset/autoaug_coco_detection.py]: done

File: easycv/datasets/loader/build_loader.py [hunk 17]
```diff
@@ -58,9 +59,10 @@ def build_dataloader(dataset,
     Returns:
         DataLoader: A PyTorch dataloader.
     """
+    seed = sync_random_seed(seed)
```
Comment (wenmengzhou) [inline easycv/datasets/loader/build_loader.py:62]: move to line 66, should be executed only when dist is True

File: easycv/datasets/loader/build_loader.py [hunk 18]
```diff
@@ -58,9 +59,10 @@ def build_dataloader(dataset,
...
+    rank, world_size = get_dist_info()
```
Comment (wenmengzhou) [inline easycv/datasets/loader/build_loader.py:63]: move to line 66, should be executed only when dist is True

Comment (tuofeilunhifi) [reply easycv/datasets/loader/build_loader.py]: done

File: easycv/datasets/loader/build_loader.py [hunk 19]
```diff
@@ -145,12 +153,13 @@ def build_dataloader(dataset,
     return data_loader
 
 
-def worker_init_fn(worker_id, seed=None, odps_config=None):
-    if seed is not None:
-        worker_seed = worker_id + seed
-        np.random.seed(worker_seed)
-        random.seed(worker_seed)
-        torch.manual_seed(worker_seed)
+def worker_init_fn(worker_id, num_workers, rank, seed, odps_config=None):
```
Comment (wenmengzhou) [inline easycv/datasets/loader/build_loader.py:156]: add support when seed is None

File: easycv/datasets/loader/sampler.py [hunk 20]
```diff
@@ -218,7 +220,6 @@ def generate_new_list(self):
     def set_uniform_indices(self, labels, num_classes):
         self.unif_sampling_flag = True
         assert self.shuffle, 'Using uniform sampling, the indices must be shuffled.'
-        np.random.seed(self.epoch)
```
Comment (wenmengzhou) [inline easycv/datasets/loader/sampler.py:221]: why remove？

File: easycv/datasets/loader/sampler.py [hunk 21]
```diff
@@ -161,6 +161,7 @@ def __init__(
         num_replicas=None,
         rank=None,
         shuffle=True,
+        seed=0,
```
Comment (wenmengzhou) [inline easycv/datasets/loader/sampler.py:164]: update doc

File: easycv/models/loss/__init__.py [hunk 22]
```diff
@@ -4,3 +4,4 @@
 from .iou_loss import GIoULoss, IoULoss, YOLOX_IOULoss
 from .mse_loss import JointsMSELoss
 from .pytorch_metric_learning import *
+from .set_criterion import *
```
Comment (wenmengzhou) [inline easycv/models/loss/__init__.py:7]: use explict import rather than import *

File: easycv/datasets/loader/build_loader.py (hunk 19 above)
Comment (tuofeilunhifi) [reply easycv/datasets/loader/build_loader.py:156]: Line 104 is already considered

File: easycv/datasets/loader/sampler.py (hunk 20 above)
Comment (tuofeilunhifi) [reply easycv/datasets/loader/sampler.py:221]: Seed should be managed uniformly and should not be abused

File: easycv/models/loss/__init__.py (hunk 22 above)
Comment (tuofeilunhifi) [reply easycv/models/loss/__init__.py]: done

File: easycv/datasets/loader/sampler.py (hunk 21 above)
Comment (tuofeilunhifi) [reply easycv/datasets/loader/sampler.py:164]: done

File: easycv/models/loss/focal_loss.py [hunk 23]
```diff
@@ -151,8 +151,8 @@ def forward(self,
 
 
 # This method is only for debugging
-def py_sigmoid_focal_loss(pred,
-                          target,
+def py_sigmoid_focal_loss(inputs,
```
[the remark this answers is not in the document]
Comment (tuofeilunhifi) [reply easycv/models/loss/focal_loss.py:153]: done

**Review by wenmengzhou — APPROVED** [review]
LGTM
